File: medical_ws1.0-master/medical_ws1.0-master/src/pid_tracking/scripts/pid.py
class PID:
    def __init__(self, mode, PID_values, max_out, max_iout):
        self.mode = mode
        self.Kp = PID_values[0]
        self.Ki = PID_values[1]
        self.Kd = PID_values[2]
        self.max_out = max_out
        self.max_iout = max_iout
        self.Dbuf = [0.0, 0.0, 0.0]
        self.error = [0.0, 0.0, 0.0]
        self.Pout = 0.0
        self.Iout = 0.0
        self.Dout = 0.0
        self.out = 0.0
        self.set = 0.0
        self.fdb = 0.0

    def init(self, mode, PID_values, max_out, max_iout):
        self.mode = mode
        self.Kp = PID_values[0]
        self.Ki = PID_values[1]
        self.Kd = PID_values[2]
        self.max_out = max_out
        self.max_iout = max_iout
        self.Dbuf = [0.0, 0.0, 0.0]
        self.error = [0.0, 0.0, 0.0]

    def limit_max(self, input_val, max_val):
        if input_val > max_val:
            input_val = max_val
        elif input_val < -max_val:
            input_val = -max_val
        return input_val

    def calc(self, ref, set_val):
        self.error[2] = self.error[1]
        self.error[1] = self.error[0]
        self.set = set_val
        self.fdb = ref
        self.error[0] = set_val - ref

        if self.mode == "PID_POSITION":
            self.Pout = self.Kp * self.error[0]
            self.Iout += self.Ki * self.error[0]
            self.Dbuf[2] = self.Dbuf[1]
            self.Dbuf[1] = self.Dbuf[0]
            self.Dbuf[0] = self.error[0] - self.error[1]
            self.Dout = self.Kd * self.Dbuf[0]
            self.Iout = self.limit_max(self.Iout, self.max_iout)
            self.out = self.Pout + self.Iout + self.Dout
            self.out = self.limit_max(self.out, self.max_out)

        elif self.mode == "PID_DELTA":
            self.Pout = self.Kp * (self.error[0] - self.error[1])
            self.Iout = self.Ki * self.error[0]
            self.Dbuf[2] = self.Dbuf[1]
            self.Dbuf[1] = self.Dbuf[0]
            self.Dbuf[0] = self.error[0] - 2.0 * self.error[1] + self.error[2]
            self.Dout = self.Kd * self.Dbuf[0]
            self.out += self.Pout + self.Iout + self.Dout
            self.out = self.limit_max(self.out, self.max_out)

        return self.out

    def clear(self):
        self.error = [0.0, 0.0, 0.0]
        self.Dbuf = [0.0, 0.0, 0.0]
        self.out = self.Pout = self.Iout = self.Dout = 0.0
        self.fdb = self.set = 0.0

File: medical_ws1.0-master/medical_ws1.0-master/src/pid_tracking/src/pid.cpp
#include "pid_tracking/pid.h"

#define LimitMax(input, max)   \
    {                          \
        if (input > max)       \
            input = max;       \
        else if (input < -max) \
            input = -max;      \
    }
PID::PID(){

}
/**
 * @brief Construct a new PID object
 *
 * @param mode
 * @param PID
 * @param max_out
 * @param max_iout
 */
void PID::init(uint8_t mode, const float PID[3], float max_out, float max_iout)
{
    this->mode = mode;
    this->Kp = PID[0];
    this->Ki = PID[1];
    this->Kd = PID[2];
    this->max_out = max_out;
    this->max_iout = max_iout;
    std::fill(std::begin(Dbuf), std::end(Dbuf), 0.0f);
    std::fill(std::begin(error), std::end(error), 0.0f);
    Pout = Iout = Dout = out = 0.0f;
}
/**
 * @brief 计算
 *
 * @param ref
 * @param set
 * @return float
 */
float PID::calc(float ref, float set)
{
    error[2] = error[1];
    error[1] = error[0];
    this->set = set;
    fdb = ref;
    error[0] = set - ref;

    if (mode == PID_POSITION)
    {
        Pout = Kp * error[0];
        Iout += Ki * error[0];
        Dbuf[2] = Dbuf[1];
        Dbuf[1] = Dbuf[0];
        Dbuf[0] = (error[0] - error[1]);
        Dout = Kd * Dbuf[0];
        LimitMax(Iout, max_iout);
        out = Pout + Iout + Dout;
        LimitMax(out, max_out);
    }
    else if (mode == PID_DELTA)
    {
        Pout = Kp * (error[0] - error[1]);
        Iout = Ki * error[0];
        Dbuf[2] = Dbuf[1];
        Dbuf[1] = Dbuf[0];
        Dbuf[0] = (error[0] - 2.0f * error[1] + error[2]);
        Dout = Kd * Dbuf[0];
        out += Pout + Iout + Dout;
        LimitMax(out, max_out);
    }
    return out;
}
/**
 * @brief 清空累积误差
 *
 */
void PID::clear()
{
    std::fill(std::begin(error), std::end(error), 0.0f);
    std::fill(std::begin(Dbuf), std::end(Dbuf), 0.0f);
    out = Pout = Iout = Dout = 0.0f;
    fdb = set = 0.0f;
}
PID::~PID()
{
    clear();
}